### ISSUE TITLE
[A11y] Search Results - Grid - "CREATE NEW COLLECTION" modal

### ISSUE THREAD

**Issue by jechols** [OregonDigital/OD2#1949]
### Operating System

Windows

### Browser

Edge

### Affected page

Search

### Admin user?

Yes

### Issue Description

This is going to be an awful mega-ticket.  This modal is difficult enough to get to that it almost needs to be audited as its own page.

#### General modal issues

These are similar to other issues we have with OD's modals:

- Tabbing forward hits an empty element for some reason before it wraps back to the first focusable element
- Tabbing backward, focus just gets "stuck" and you can only tab forward to get back to the modal.
- The close button on this modal is intermittently visible to NVDA. In #1931 I described the close button being invisible, but I think it changes state in javascript somehow. The first time I tabbed to it, it announced "times button", but since then it's been "blank" and cannot be activated.
- The dialog isn't coded proper to announce itself as a dialog. In the ARIA practices example, you hear the dialog title and that it's a dialog when you open it: https://www.w3.org/TR/wai-aria-practices-1.1/examples/dialog-modal/dialog.html

##### On adding a new title when the box is empty

- The region which displays the error ("cannot add another with empty field") isn't announced to assistive tech. There's an aria-live wrapping the area, but:
  -  It's set to polite which is wrong for an error users need to know about
  - It wraps the whole area, field and all, instead of just the region you'd want announced. This could be a problem for making it work properly.
  - It doesn't work. It may be because it wraps too many irrelevant elements or maybe the DOM issue (below) makes it break.
- The DOM gets broken somehow via whatever provides this message: there's a `div` inside a `ul` element

##### General

- The "ADD ANOTHER TITLE" button announces the plus - that symbol should be hidden.
  - The "plus" symbol is added via a CSS glyphicon span element. This probably seemed like exactly the right way to add something so a screen reader wouldn't announce it. Unfortunately, SRs do a lot of guesswork. The recommendation I've read is to assume CSS-created elements/content will be read when you want it silent, and it will be silent when you want it read. So if you use CSS for adding content like this, you still always want to explicitly ensure the accessible name/label contains only the elements you mean for it to. In this case, the span with the glyphicon should be hidden to assistive tech entirely.
- When there's a second title field, it has no label
- All title fields share the same `id` attribute
- The "help" text isn't associated with the fields. e.g., assistive tech users have a very difficult time finding "A name to aid in identifying a collection" when they're in the "Title" field.
- When no title is in the form, submitting the form closes the dialog with no explanation. This is a usability issue more than accessibility, but a sighted user can at least see that the dialog is gone.
- Activating the "Remove" button doesn't announce that anything happened. There should be some kind of announcement (and probably a visual indicator) that a title was in fact removed.
- The "required" label near "Title" has very poor color contrast
- The "REMOVE" button's color contrast is slightly too low
- "CREATE NEW COLLECTION" functions as a heading but isn't coded as such

### Suggested Fix

General modal issues have been described in other tickets, but basically everything described at https://www.w3.org/TR/wai-aria-practices-1.1/#dialog_modal needs to be implemented.

- Figure out what's making the `div` show up inside a `ul` and correct that
- When a title can't be added, make sure the aria-live region is updated, and doesn't wrap so many elements that assistive tech might have trouble catching the updates.
  - Preferably the aria-live region should only wrap the part that will contain a dynamic message, definitely not the whole form.
  - Make sure `aria-live` is set to `assertive`, not `polite`. Polite is meant for messages that aren't critical, but error messages should always be announced ASAP.
  - An "alert" region may make sense to avoid manually setting `aria-live`: https://www.w3.org/TR/wai-aria-practices-1.1/examples/alert/alert.html
- Make sure the "plus" symbol near "ADD ANOTHER TITLE" is hidden from screen readers
- Make sure all fields always have a label, even when they're dynamically created fields. The attribute [aria-labelledby](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-labelledby_attribute) may be useful if the label should always be "Title", otherwise `aria-label` could just be set if the label should be something custom like "Second title", "Third title", etc.  The latter is always more risky, though, because it's easier for mistakes to go unnoticed when a label is completely hidden inside an ARIA attribute.
- Use `aria-describedby` to associate help text with fields
- When duplicating the title fields make sure unique ids are set up (if they need ids)
- Add an indicator when titles are removed
- Change contrast for the "required" label
- Make the "REMOVE" button's a slightly darker red
- Set "CREATE NEW COLLECTION" to be a level two heading

### Can anybody QA or does this need an expert?

Needs expert

### QA Steps

Not sure there are steps here. The modal will probably need a complete re-audit because this one is particularly complex from an a11y standpoint.

**Comment by jechols** [comment]
Problems which still exist:

The region that tells the user "cannot add another with empty field" is still broken. The aria-live region doesn't exist until the error occurs, and I believe this makes assistive tech unable to see it. I believe the region needs to be in the dom and registered by the AT prior to a change so AT knows to watch for changes.

When the dialog opens, I think focus is being managed incorrectly somehow. When the dialog opens, NVDA reads out information about buttons that aren't on it (select all / none buttons), then starts reading the dialog from top to bottom. This is very confusing. The aria example on https://www.w3.org/TR/wai-aria-practices-1.1/examples/dialog-modal/dialog.html, NVDA simply says, "Add Delivery Address dialog Street has autocomplete blank" - letting the user know they're in the dialog and on a form field called "Street".

When you add a title, its "id" attribute is the same as the first title field.

The "help" text isn't associated with the "Description" field. The "Title" field has been fixed, but not "Description".

The form still closes on "save" even when it's not valid, and with no information about it having done so. Probably not a hard failure, but definitely problematic that nothing is announced via assistive tech.

Activating the "Remove" button doesn't announce that anything happened. There should be some kind of announcement (and probably a visual indicator) that a title was in fact removed.
